### PR TITLE
Python: ADR for new AI connector abstract methods

### PR DESCRIPTION
**Pull request by TaoChenOSU**
### Motivation and Context

This ADR documents a proposal to introduce four new abstract methods in `ChatCompletionClientBase` and `TextCompletionClientBase` to optimize code and further streamline the process of adding new AI connectors.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: docs/decisions/0052-python-ai-connector-new-abstract-methods.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+---
+# These are optional elements. Feel free to remove any of them.
+status: { proposed }
+contact: { Tao Chen }
+date: { 2024-08-30 }
+deciders: {}
+consulted: {}
+informed: {}
```
Comment (TaoChenOSU) [inline docs/decisions/0052-python-ai-connector-new-abstract-methods.md:8]: Will edit this section once the ADR is approved.

**Comment by nmoeller** [comment]
@TaoChenOSU i really like this ! With this we could potentially handle the Function Calls in `ChatCompletionClientBase` Class, this would remove a lot of duplicate Code in the Connectors !

**Review by eavanvalkenburg — APPROVED** [review]
This is great, Tao, thanks for writing this up, I like the thinking, I've used a very similar setup for the new memory connectors, where the classes implement a `_inner_*` method, while the `get`, `upsert`, etc methods are all default implemented in the base class, indeed makes it much simpler to implement a new one! One thing to add is a way to indicate which model can and can't do function calling, since the function call behavior is part of the base prompt execution settings and that is therefore not a way to distinguish!

**Comment by TaoChenOSU** [comment]
> This is great, Tao, thanks for writing this up, I like the thinking, I've used a very similar setup for the new memory connectors, where the classes implement a `_inner_*` method, while the `get`, `upsert`, etc methods are all default implemented in the base class, indeed makes it much simpler to implement a new one! One thing to add is a way to indicate which model can and can't do function calling, since the function call behavior is part of the base prompt execution settings and that is therefore not a way to distinguish!

That's great! I think a `ClassVar` will be sufficient to indicate whether a connector will support function calling. I will add that.

File: docs/decisions/0052-python-ai-connector-new-abstract-methods.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+### A new `ClassVar[bool]` variable in `ChatCompletionClientBase` to indicate whether a connector supports function calling
+
+This class variable will be overridden in derived classes and be used in the default implementations of `get_chat_message_contents` and `get_streaming_chat_message_contents`.
+
+```python
+class ChatCompletionClientBase(AIServiceClientBase, ABC):
+    """Base class for chat completion AI services."""
+
+    FUNCTION_CALLING_SUPPORT: ClassVar[bool] = False
```
Comment (alliscode) [inline docs/decisions/0052-python-ai-connector-new-abstract-methods.md:61]: nit: name it as a question that's answered by the bool e.g. SUPPORTS_FUNCTION_CALLING, HAS_FUNCTION_CALLING, etc.

**Review by alliscode — APPROVED** [review]

**Review by alliscode — APPROVED** [review]

**Pull request merged**